src/components/synapse.rs

```diff
@@ -45,22 +54,52 @@ impl SynapseComponent {
         result
     }
 
-    pub async fn who_am_i(&self, token: &str) -> Result<WhoAmIResponse, String> {
+    pub async fn who_am_i(&self, token: &str) -> Result<WhoAmIResponse, CommonError> {
         let who_am_i_url = format!("{}{}", self.synapse_url, WHO_AM_I_URI);
         let client = reqwest::Client::new();
-        let result: Result<WhoAmIResponse, String> = match client
+        match client
             .get(who_am_i_url)
-            .header("authorization", token)
+            .header("Authorization", format!("Bearer {}", token))
             .send()
             .await
         {
-            Ok(response) => match response.json::<WhoAmIResponse>().await {
-                Ok(json) => Ok(json),
-                Err(err) => Err(err.to_string()),
-            },
-            Err(err) => Err(err.to_string()),
-        };
+            Ok(response) => {
+                let text = response.text().await;
+                if let Err(err) = text {
+                    log::warn!("error reading synapse response {}", err);
+                    return Err(CommonError::Unknown);
+                }
 
-        result
+                let text = text.unwrap();
+
+                let who_am_i_response = serde_json::from_str::<WhoAmIResponse>(&text);
+
+                match who_am_i_response {
+                    Ok(who_am_i) => Ok(who_am_i),
+                    Err(_) => Err(SynapseComponent::parse_and_return_error(&text)),
+                }
+            }
+            Err(err) => {
+                log::warn!("error connecting to synapse {}", err);
+                Err(CommonError::Unknown)
+            }
+        }
+    }
+
+    fn parse_and_return_error(text: &str) -> CommonError {
+        let error_response = serde_json::from_str::<SynapseErrorResponse>(&text);
+
+        match error_response {
+            Ok(error) => match error.errcode.as_str() {
+                "M_UNKNOWN_TOKEN" => CommonError::Unauthorized,
+                "M_MISSING_TOKEN" => CommonError::Unauthorized,
+                "M_LIMIT_EXCEEDED" => CommonError::TooManyRequests,
+                _ => CommonError::Unknown,
+            },
+            Err(err) => {
+                log::warn!("error parsing synapse error {}", err);
```

Comment:
why not `log::error`?

Reply:
I'd keep `log::error` for OUR errors, this would be a error from synapse